### ISSUE TITLE
convert a model(trained by lit-gpt) to AutoModelForCausalLM model

### ISSUE THREAD

**Issue by pull-ups** [Lightning-AI/lit-gpt#844]
Hi, thanks for providing easy-to-do program for tuning language model. 
But I found that the trained model is saved as "checkpoint.pth" file, and model's class type is "GPT". Is there any way that I can convert this model into AutoModelForCausalLM, which is frequently used for any purpose?

**Comment by carmocca** [comment]
Check out this tutorial: https://github.com/Lightning-AI/lit-gpt/blob/main/tutorials/convert_lit_models.md

**Comment by pull-ups** [comment]
Thanks. but can you explain what path I should designate for config_path? I don't have a config.json file in checkpoint directory and lora merged directory. (after lora training, I merged lora weight into backbone model by using "scripts/merge_lora.py")

In the directory of checkpoint_path, I got 
* generation_config.json
* pytorch_model-00001-of-00002.bin
* pytorch_model-00002-of-00002.bin
* tokenizer_config.json
* tokenizer.json
* pytorch_model.bin.index.json


in lora merged directory, I got
* lit_model.pth

python scripts/convert_lit_checkpoint.py \
    --checkpoint_path path/to/litgpt/model.pth \
    --output_path where/to/save/the/converted.ckpt \
    **--config_path path/to/litgpt/config.json**
    
    where can I find config.json?

**Comment by Andrei-Aksionov** [comment]
Hello @pull-ups 

After you converted the downloaded weights by `scripts/convert_hf_checkpoint.py` it created this config file.
Most likely, you accidentally deleted it.
Otherwise, the merge script wouldn't work, since it requires this config too:
https://github.com/Lightning-AI/lit-gpt/blob/70164f834f38e72523b92424a63c3bf20006fd40/scripts/merge_lora.py#L51-L52

**Comment by PIXPOINT** [comment]
Hi,
I'm also in the process of exporting the model created with lit-gpt. This also worked with `scripts/convert_hf_checkpoint.py`. I then tried `llama.cpp python3 convert.py /path/to/model/model.ckpt --ctx 4096` to get it running in LM Studio. Unfortunately that doesn't work because the file cannot be read. Does anyone have a tip or idea how and where the .ckpt can be executed?

**Comment by Andrei-Aksionov** [comment]
Just tried to convert and run model in `llama.cpp` and all went fine (though required a bit of tinkering).
Maybe you missed a step or two?

So I did LoRA fine-tune with TinyLLaMa model and the code from the current main. The only change I've made is reduced the number of max_iter to 100 (for the sake of speed).

Here are the steps that I've made:
```bash
# 1. Export the name of the model for convinience
export model=TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T 

# 2. Download the weights
python scripts/download.py --repo_id $model

# 3. Convert HF weight to Lit format
python scripts/convert_hf_checkpoint.py --checkpoint_dir checkpoints/$model

# 4. Prepare data. For this example I used Alpaca
python scripts/prepare_alpaca.py --checkpoint_dir checkpoints/$model

# 5. LoRA fine-tuning
python finetune/lora.py --checkpoint_dir checkpoints/$model

# 6. Merge pretrained and LoRA weights
python scripts/merge_lora.py --checkpoint_dir checkpoints/$model --lora_path out/lora/alpaca/lit_model_lora_finetuned.pth --out_dir out/lora/alpaca/merged_ckpt

# 7. Convert the merged model in Lit format back to HF format
python scripts/convert_lit_checkpoint.py --checkpoint_path out/lora/alpaca/merged_ckpt/lit_model.pth --output_path out/lora/alpaca/converted.ckpt --config_path checkpoints/$model/lit_config.json

# Now the model is in the HF format.
# In order to convert the model with the help of Llama.cpp you need to copy the converted model 
# back to checkpoint folder (in this case it's `checkpoints/$model`) so Llama.cpp can access 
# required files like tokenizer and also you have to download additional file (`config.json`) 
# from the HF model page and put it in the same folder as the converted weights. 
# Code snippets for these steps are #8 and #9

# 8
cp out/lora/alpaca/converted.ckpt checkpoints/$model

# 9
wget https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-955k-token-2T/raw/main/config.json -P checkpoints/$model
```

Now you can execute `python convert.py [model_path] --ctx 2048` from `llama.cpp`.

**Comment by pull-ups** [comment]
@Andrei-Aksionov  Thank you for your explanation! I've encountered another problem. 

I want to use the fine-tuned model as normal case(by AutoModelForCasualLM), not as GPT class( used in lit-gpt), because I want to apply this fine-tuned model to DPO code that I have. That DPO code is for AutoModelForCasualLM.
So, I loaded fine-tuned model weight via below code:
`
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

base_model_name = "checkpoints/mistralai/Mistral-7B-Instruct-v0.2" (downloaded from huggingface)
model = AutoModelForCausalLM.from_pretrained(base_model_name,  device_map='auto')

checkpoint_file='out/converted/Mistral_beethoven2/converted.ckpt' (output file after running convert_lit_checkpoint.py)
model.load_state_dict(torch.load(checkpoint_file))
`

But, after running inference with same input and same tokenizer, I got different generation output. It's not broken at all, but slightly different with the result that I get with running "generate/base.py". My guess is, the different is from generation configs. So, I tried to set settings same as lit-gpt's generation and my code(AutoModelForCasualLM) generation.

in generate/base.py, 
`y = generate(model, encoded, max_returned_tokens, temperature=0.8, top_k=200, eos_id=tokenizer.eos_id)`
in my code (generation part), 
`generated_ids = model.generate(model_inputs, top_k=200, temperature=0.8, pad_token_id=tokenizer.eos_id)`

Is there any way that loading fine-tuned weight(converted.ckpt) into AutomodelforCasualLM? 
Thanks.

**Comment by Andrei-Aksionov** [comment]
I might be wrong, but the parity in generation output between Lit-GPT and Transformers is not a goal. So if there is a small deviation in the generated output, it might not be an issue. @carmocca could you confirm that?

> Is there any way that loading fine-tuned weight(converted.ckpt) into AutomodelforCasualLM?

That bit I didn't understand. Could you perhaps rephrase it?

**Comment by pull-ups** [comment]
Thank for your reply. I mean, 
in this library(lit-gpt)'s train code and inference code, all of the models are called as class "GPT" (defined in lit-gpt/lit_gpt/model.py) and tokenizers are called as class "Tokenizer" (defined in lit_gpt/lit_gpt/tokenizer.py) But, many nowadays LLMs (like Mistral) use AutoModelForCausalLM, AutoTokenizer. (https://huggingface.co/docs/transformers/main/en/model_doc/mistral) for their normal use. 

I tried to load fine-tuned model's weight via below code:

```
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

base_model_name = "checkpoints/mistralai/Mistral-7B-Instruct-v0.2" (downloaded from huggingface)
model = AutoModelForCausalLM.from_pretrained(base_model_name,  device_map='auto')
checkpoint_file='out/converted/Mistral_beethoven2/converted.ckpt' (output file after running convert_lit_checkpoint.py)
model.load_state_dict(torch.load(checkpoint_file))
```

Is this correct method for loading fine-tuned model weight (fine-tuned with lit-gpt) into AutoModelForCausalLM? or is there any other way?

I'm online right now, so If you don't get what I said, please tell me. Thank you for your help.

**Comment by Andrei-Aksionov** [comment]
> Is this correct method for loading fine-tuned model weight (fine-tuned with lit-gpt) into AutoModelForCausalLM?

Otherwise it wouldn't work :)

If you take a look at the output of `torch.load(checkpoint_file)` you'll see that this is a dictionary, where `key` is the name of the parameter that reflects the structure of the model, like `lm_head.weight` or `model.layers.0.input_layernorm.weight` (for TinyLlaMa). There is no mentioning of `AutoModelForCausalLM` or anything like that.

When transformers library creates an instance of class AutoModelForCausalLM, that instance has attributes like I mentioned above.
So `model.load_state_dict(...)` just tries to match keys in that dictionary with class attribute names.

In other words, it doesn't matter what is the name of the class.
We just need that, for instance, key `lm_head_weight` in the state_dict (output of `torch.load(checkpoint_file)`) matches attribute name `self.lm_head.weight` of the class AutoModelForCausalLM.

**Comment by pull-ups** [comment]
I understand that the output file after running "convert_lit_checkpoint.py" can be applied with any type of model, (not only for AutoModelforCausalLM that I repeatedly said). 
Then, I think the parity between Lit-GPT and transformer should be from the difference of generation process. 
in Lit-GPT, generation is done by lit-gpt/generate/base.py/"generate" function, but the generation is based on the model class "GPT" of lit-gpt, not transformer.
By using transformer, How can I set same environment (or setting) as lit-gpt's generation, so that generated output from two library become same?

**Comment by Andrei-Aksionov** [comment]
> By using transformer, How can I set same environment (or setting) as lit-gpt's generation, so that generated output from two library become same?

I have no answer to this question.
It can be many things: different generation config, different generation code, it can be even a different fixed random seed, ...
Since the parity in generation output between Lit-GPT and Transformers wasn't the goal (again, @carmocca please confirm or reject my assumption), these differences were not discovered.

**Comment by carmocca** [comment]
We test for numerical parity of the model using float32, but parity of the generation itself is not a goal.

I once debugged a similar issue and found it to be equal if equivalent settings are used for each library: https://github.com/Lightning-AI/lit-gpt/issues/580#issuecomment-1733966255
